### ISSUE TITLE
Proof of Java 14 syntax by new CI build item

### ISSUE THREAD

**Issue by nmancus1** [checkstyle/checkstyle#8631]
We need to prove that our java 14 grammar can parse all compilable code in https://github.com/openjdk/jdk14u.  To do this, we need to create a new build item in CI (probably Travis) that prints ASTs for all compilable code in the jdk14 repository. But, due to the many non-compilable files, we must find a solution to parse only those that are compilable. Here are some ideas:

### Include List
_________________

**This is not an ideal solution, because we would be missing thousands of files that we should be testing on**.

I have found a list of test files from the `langtools/` directory here: https://download.java.net/openjdk/testresults/15/archives/24/langtools-24-summary.txt

While this is not a comprehensive list, after removing the non-compilable files, there are 2,543 inputs that we can test on.

That list of files is here: https://gist.github.com/nmancus1/9c3be05a00278774d7879d8f5b6578fb

### Exclude List
______________________

**It would be better to exclude those files that are non-compilable and parse all compilable code in JDK14. Here are some ways that we can do that:**

We can exclude files with the `@compile/fail/ref=` javadoc tag. Here is a list of all files containing the `@compile/fail/ref=` tag. 

- List here: https://gist.github.com/nmancus1/8a4be76cb733691e0a6ef7096c3069e6

 - Shown here: 
https://github.com/openjdk/jdk/blob/b5d775f115e4274d8f0c221a8eb4a92d17a9f579/test/langtools/tools/javac/failover/FailOver15.java#L6

Another way to find files that should be excluded is to grep files with `compiler.err` comments.

 - List here: https://gist.github.com/nmancus1/155c860be9c130b69cf74e384ae5979f

 - Shown here:
https://github.com/openjdk/jdk14/blob/abc56193174dd6a11a8453bd5036fc8618ecca61/test/langtools/tools/javac/diags/examples/RecordsCantDeclareComponentModifiers.java#L24

We can build an excludes file from these lists.

**Comment by romani** [comment]
Yes, from all files, let's exclude by this tag, then exclude any other files if they problematic. At the end execute CLI tree print on exact list of files

**Comment by nmancus1** [comment]
I think we will need to print AST to really prove that we can parse all java files in jdk14:

```
➜  src cat RecordsCantDeclareComponentModifiers.java 

/*
 * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

// key: compiler.err.record.cant.declare.field.modifiers
// key: compiler.note.preview.filename
// key: compiler.note.preview.recompile
// options: --enable-preview -source ${jdk.version}

record R(final int x) {
}
                                                                                                                                                                                                                                                                       ➜  src /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 RecordsCantDeclareComponentModifiers.java                       
RecordsCantDeclareComponentModifiers.java:30: error: record components cannot have modifiers
record R(final int x) {
         ^
Note: RecordsCantDeclareComponentModifiers.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
1 error

➜  src java $RUN_LOCALE -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.36-SNAPSHOT-all.jar -c config.xml RecordsCantDeclareComponentModifiers.java
Starting audit...
Audit done.

➜  src java $RUN_LOCALE -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.36-SNAPSHOT-all.jar -t RecordsCantDeclareComponentModifiers.java           
com.puppycrawl.tools.checkstyle.api.CheckstyleException: IllegalStateException occurred while parsing file /home/nick/Desktop/full-record-grammar/check-updates/src/RecordsCantDeclareComponentModifiers.java.
        at com.puppycrawl.tools.checkstyle.JavaParser.parse(JavaParser.java:120)
        at com.puppycrawl.tools.checkstyle.JavaParser.parseFileText(JavaParser.java:137)
        at com.puppycrawl.tools.checkstyle.JavaParser.parseFile(JavaParser.java:157)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printFileAst(AstTreeStringPrinter.java:65)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:284)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
Caused by: java.lang.IllegalStateException: /home/nick/Desktop/full-record-grammar/check-updates/src/RecordsCantDeclareComponentModifiers.java:30:10: unexpected token: final
        at com.puppycrawl.tools.checkstyle.JavaParser$1.reportError(JavaParser.java:108)
        at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.typeDefinition(GeneratedJavaRecognizer.java:424)
        at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.compilationUnit(GeneratedJavaRecognizer.java:212)
        at com.puppycrawl.tools.checkstyle.JavaParser.parse(JavaParser.java:114)
        ... 6 more
Caused by: /home/nick/Desktop/full-record-grammar/check-updates/src/RecordsCantDeclareComponentModifiers.java:30:10: unexpected token: final
        at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.recordComponents(GeneratedJavaRecognizer.java:2861)
        at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.recordComponentsList(GeneratedJavaRecognizer.java:2642)
        at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.recordDefinition(GeneratedJavaRecognizer.java:818)
        at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.typeDefinitionInternal(GeneratedJavaRecognizer.java:589)
        at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.typeDefinition(GeneratedJavaRecognizer.java:402)
        ... 8 more
Checkstyle ends with 1 errors.

➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
    </module>
</module>
  
```

**Comment by nmancus1** [comment]
Files in langtools/ that causes exception when printing AST: https://gist.github.com/nmancus1/ee581a3b946a173db569359a7d05fa83

**Comment by nmancus1** [comment]
Left: grepped results of jdk file in `langtools/` with either `@compile/fail/ref= tag` or `compiler.err` comments. 
Right: files in `langtools/` that Checkstyle failed to parse.

https://www.diffchecker.com/DMPzMmxw

**Comment by pbludov** [comment]
./test/langtools/tools/javac/enum/TrailingComma.java
This is one of the non-obvious features of JLS. From the [specs](https://docs.oracle.com/javase/specs/jls/se14/html/jls-8.html#jls-8.9.1):
```
EnumBody:
{ [EnumConstantList] [,] [EnumBodyDeclarations] } 
```
The `EnumConstantList` is optional, thus this is legal:
```java
public enum Empty {
  ,
}
```
and even this:
```java
public enum Empty {
  ,
  ;
}
```
Our parser is more strict in [this case](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g#L749):
```
// This is the body of an enum. You can have zero or more enum constants
// followed by any number of fields like a regular class
enumBlock
    :    LCURLY
            ( enumConstant ( options{greedy=true;}: COMMA enumConstant )* ( COMMA )? )?
            ( SEMI ( field | SEMI )* )?
        RCURLY
        {#enumBlock = #([OBJBLOCK, "OBJBLOCK"], #enumBlock);}
    ;
```
Technically, there is an error in our parser. But since such enums do not make any sense, nobody complains yet.

**Comment by romani** [comment]
fix is merged
